text/0000-float-semantics.md

```diff
@@ -0,0 +1,272 @@
+
+Another alternative to handling floating-point operations in `const` is to just fail when a non-deterministic choice would occur at runtime, i.e., each time a NaN is produced.
+However, this is a breaking change: `const C: f32 = 0.0 / 0.0;` has worked on stable Rust since Rust 1.0.
+
+Even under this alternative we would allow `const fn` to perform operations that are non-deterministic at run-time, i.e., unsafe code could still not rely on `const fn` as being a marker for "this is deterministic at run-time". Thus it shares all the downsides with the previous alternative, it just avoids making NaN choices observable in `const`.
+
+The core advantage of this option is that it avoids having `const` results change when the unspecified compile-time NaN changes on a compiler update or across compilers.
+However, having `const` results depend on NaN bits should be very rare, and we already have other (more common) cases of `const` results depending on unspecified implementation details that can and sometimes do change on compiler updates, namely the layout of `repr(Rust)` types (observable via `size_of` and `offset_of`).
+
+### Alternative: `const` tracks NaN values, fails when their bits matter during compile-time
+
+`const` could in principle track NaNs symbolically, similar to how it tracks pointers, and delay choosing NaN payload bits until codegen.
+Const-evaluation would then abort if the bits of a NaN are observed (eg. if `to_bits`is called during const-evaluation).
+This would keep `const C = 0.0/0.0;` working, but requires `const fn is_nan` to be an intrinsic.
+However, it would require massive amounts of work in the compile-time interpreter, comparable in complexity to all the work that is already required to support symbolic pointers (and the RFC author doubts that there will be a lot of opportunity for those two kinds of symbolic state to share infrastructure).
+That effort should only be invested if there is a significant payoff.
+The RFC author considers the downsides of unspecified NaN bit patterns being observable in const-evaluation to be minimal, and hence the payoff of this alternative to be low.
+
+# Prior art
+[prior-art]: #prior-art
+
+C and C++ generally avoid saying anything about signaling NaNs, and also don't talk about how NaNs are chosen. I can only assume this means they are chosen non-deterministically; that is certainly the interpretation chosen by GCC and LLVM.
+
+GCC [says](https://gcc.gnu.org/wiki/FloatingPointMath) "Without any explicit options, GCC assumes round to nearest or even and does not care about signalling NaNs". It is unclear whether "does not care" also means "guarantees to never produce by itself", i.e. whether `0.0 / 0.0` is ever allowed to evaluate to a signaling NaN or not. If it *is* allowed to evaluate to a signaling NaN, that is probably a violation of the C standard, which guarantees that `pow(1, 0.0/0.0)` returns `1` -- but in practice, `pow(1, sNaN)` returns a NaN.
```

Comment:
I don't have firm records of the history of CFP deliberation on the wording of C §F.2.1. IEEE 754 makes it clear that sNaNs are not producible by regular operations under normal situations, and the footnote attached to the statements quoted above in §F.2.1 also makes it clear that nonsupport for sNaN isn't an issue:

> Since NaN s created by IEC 60559 arithmetic operations are always quiet, quiet NaNs (along with infinities) are sufficient for closure of the arithmetic.

`0.0 / 0.0` can only produce a qNaN, even in don't-care-about-sNaN mode.

Reply:
Yeah, under IEEE 754 there certainly can't be any sNaN produced by arithmetic operations.

But under optimizing compilers, sNaN *can* be produced by arithmetic operations, so looking at the IEEE spec won't help us here. We have to look at the compiler docs and the specs of the language they implement. For LLVM we now have very clear docs telling us when arithmetic can produce an sNaN (in particular, `0.0 / 0.0` can *not* produce an sNaN). I am not aware of any such documentation for GCC.